src/operator/contrib/transformer.cu

```diff
@@ -22,12 +22,898 @@
+    const int32_t output_lin_kv_dim = inputs[0].shape_[2];
+    const int32_t attn_batches      = inputs[1].shape_[0];
+    const int32_t q_seq_len         = inputs[1].shape_[1];
+    const int32_t embed_dim         = output_lin_kv_dim / 2;
+    int32_t head_dim                = embed_dim / params.heads;
+    const int32_t lead_dim_kv       = attn_batches * head_dim * 2;
+    const int32_t batch_stride_kv   = 2 * head_dim;
+    const float alpha               = 1.f;
+    const float beta                = req[0] == kAddTo ? 1.f : 0.f;
+
+    if (req[0] == kNullOp)
+      return;
+
+    gemm_switch_fp32accum(s,
+                          false,
+                          false,
+                          head_dim,
+                          q_seq_len,
+                          kv_seq_len,
+                          alpha,
+                          keys_values + head_dim,
+                          lead_dim_kv,
+                          batch_stride_kv,
+                          attention_maps,
+                          kv_seq_len,
+                          kv_seq_len * q_seq_len,
+                          beta,
+                          output,
+                          head_dim * attn_batches,
+                          head_dim,
+                          attn_batches);
+  })
+}
+
+void BackwardInterleavedMatMulEncDecValAttGPU(const nnvm::NodeAttrs& attrs,
+                                              const OpContext &ctx,
+                                              const std::vector<TBlob> &inputs,
+                                              const std::vector<OpReqType> &req,
+                                              const std::vector<TBlob> &outputs) {
+  const auto& params = nnvm::get<InterleavedMatMulParam>(attrs.parsed);
+  mshadow::Stream<gpu>* s = ctx.get_stream<gpu>();
+  MSHADOW_REAL_TYPE_SWITCH(inputs[0].type_flag_, DType, {
+    const DType* output_grads   = inputs[0].FlatTo2D<gpu, DType>(s).dptr_;
+    const DType* keys_values    = inputs[1].FlatTo2D<gpu, DType>(s).dptr_;
+    const DType* attention_maps = inputs[2].FlatTo2D<gpu, DType>(s).dptr_;
+    DType* keys_values_grads    = outputs[0].FlatTo2D<gpu, DType>(s).dptr_;
+    DType* attention_maps_grads = outputs[1].FlatTo2D<gpu, DType>(s).dptr_;
+    const int32_t kv_seq_len        = inputs[1].shape_[0];
+    const int32_t sequences         = inputs[1].shape_[1];
+    const int32_t output_lin_kv_dim = inputs[1].shape_[2];
+    const int32_t attn_batches      = inputs[2].shape_[0];
+    const int32_t q_seq_len         = inputs[2].shape_[1];
+    const int32_t embed_dim         = output_lin_kv_dim / 2;
+    int32_t head_dim                = embed_dim / params.heads;
+    const int32_t lead_dim_kv       = attn_batches * head_dim * 2;
+    const int32_t batch_stride_kv   = 2 * head_dim;
+    const float alpha               = 1.f;
+
+    if (req[0] != kNullOp) {
+      if (req[0] == kWriteTo && !params.bwd_ignore_zero_init) {
```

Comment:
Why should we add this flag?

Reply:
I cannot  well understand the rationale behind this flag. I think we can determine whether we need to inplace_accumulate the gradient when  `req[0] == kAddTo`

Reply:
As I put in the description, this is a hack if you use the Ops within an multihead attention the input tensor is an interlacing of projected QKV,  one Op do backward on QK part the other do backward on V (and attention weights). because of this you actually never need to "accumulate" the gradients between both Ops. It's hack that you only enable if you use this flag `bwd_ignore_zero_init`. If you intend to use this Op out of the official multihead attention don't enable this flag and the behavior will be as expected.